reasoner/resolution/ResolverRegistry.java

```diff
@@ -18,51 +18,78 @@
 import grakn.core.reasoner.resolution.resolver.RootResolver;
 import grakn.core.reasoner.resolution.resolver.RuleResolver;
+import graql.lang.pattern.variable.Reference;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.util.HashMap;
-import java.util.List;
+import java.util.Map;
+import java.util.Optional;
 import java.util.function.Consumer;
-
 public class ResolverRegistry {
 
     Logger LOG = LoggerFactory.getLogger(ResolverRegistry.class);
-    private final HashMap<Long, Actor<ConcludableResolver>> concludables;
-    private final HashMap<List<Long>, Actor<RuleResolver>> rules;
+    private final HashMap<ConjunctionConcludable<?, ?>, Actor<ConcludableResolver>> concludableActorsMap;
+    private final HashMap<Rule, Actor<RuleResolver>> rules;
     private final Actor<ResolutionRecorder> resolutionRecorder;
-    private final EventLoopGroup elg;
+    private EventLoopGroup elg;
 
     public ResolverRegistry(EventLoopGroup elg) {
         this.elg = elg;
-        concludables = new HashMap<>();
+        concludableActorsMap = new HashMap<>();
         rules = new HashMap<>();
         resolutionRecorder = Actor.create(elg, ResolutionRecorder::new);
     }
 
-    public Actor<ConcludableResolver> registerConcludable(Long pattern, List<List<Long>> rules, long traversalSize) {
-        LOG.debug("Register retrieval for concludable actor: '{}'", pattern);
-        return concludables.computeIfAbsent(pattern, (p) -> Actor.create(elg, self -> new ConcludableResolver(self, p, rules, traversalSize)));
+    public Pair<Actor<ConcludableResolver>, Map<Reference.Name, Reference.Name>> registerConcludable(ConjunctionConcludable<?, ?> concludable) {
+        LOG.debug("Register retrieval for concludable actor: '{}'", concludable.conjunction());
+
+        final Optional<Pair<ConjunctionConcludable<?, ?>, AlphaEquivalence>> alphaEquivalencePair = concludableActorsMap.keySet().stream()
+                .map(k -> new Pair<ConjunctionConcludable<?, ?>, AlphaEquivalence>(k, k.alphaEquals(concludable)))
+                .filter(p -> p.second().isValid()).findAny();
+
+        final Actor<ConcludableResolver> concludableActor;
+        final Map<Reference.Name, Reference.Name> variableMapping;
+        if (alphaEquivalencePair.isPresent()) {
+            // Then we can use the same ConcludableActor, but with a different variable mapping
+            concludableActor = concludableActorsMap.get(alphaEquivalencePair.get().first());
+            variableMapping = alphaEquivalencePair.get().second().asValid().namedVariableMapping();
+        } else {
+            // Create a new ConcludableActor
+            concludableActor = Actor.create(elg, self -> new ConcludableResolver(self, concludable));
+            concludableActorsMap.put(concludable, concludableActor);
+            variableMapping = MappingAggregator.identityMapping(concludable);
+        }
+        return new Pair<>(concludableActor, variableMapping);
```

Comment:
If we found a match, return that ConcludableActor and its VariableMapping.

Otherwise, create a new ConcludableActor, with a trivial (identity) variableMapping.